### PR TITLE
Add type declaration . Thank you rector.

### PR DESCRIPTION
**Pull request by harikt**
Using rector to make the necessary changes to code for type hint and other auto rules.

Current `rector.php`

```php
<?php

declare(strict_types=1);

use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Rector\Core\Configuration\Option;
use Rector\Set\ValueObject\SetList;
use Rector\Set\ValueObject\LevelSetList;

return static function (ContainerConfigurator $containerConfigurator): void {
    // get parameters
    $parameters = $containerConfigurator->parameters();

    $parameters->set(Option::PATHS, [
        __DIR__ . '/src'
    ]);

    $parameters->set(Option::SKIP, [
        'src/Locator/ValidateLocator.php',
        'src/Locator/SanitizeLocator.php',
    ]);

    $parameters->set(Option::AUTO_IMPORT_NAMES, true);
    $containerConfigurator->import(LevelSetList::UP_TO_PHP_72);
    $containerConfigurator->import(SetList::PHP_72);
    $containerConfigurator->import(SetList::TYPE_DECLARATION);
    $containerConfigurator->import(SetList::TYPE_DECLARATION_STRICT);
};
```

### CODE REVIEW

File: src/Rule/Validate/Email.php [hunk 1]
```diff
@@ -1142,7 +1143,7 @@ protected function parseComponentLiteral(): void
                         } elseif ((substr($IPv6, -1) === Email::STRING_COLON) && (substr($IPv6, -2, 1) !== Email::STRING_COLON)) {
                             // Address ends with a single colon
                             $this->returnStatus[] = Email::RFC5322_IPV6_COLONEND;
-                        } elseif (count(preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT)) !== 0) {
+                        } elseif ((is_array(preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT)) || preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT) instanceof Countable ? count(preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT)) : 0) !== 0) {
```
Comment (koriym) [inline src/Rule/Validate/Email.php]: preg_grep() never return Countable?
It also makes the code harder to read.

File: src/Rule/Validate/Email.php [hunk 2]
```diff
@@ -1667,11 +1668,11 @@ protected function checkDns(): void
                 // Domain can't be found in DNS
                 $this->returnStatus[] = Email::DNSWARN_NO_RECORD;
             } else {
-                if (count($result) === 0) {
+                if ((is_array($result) || $result instanceof Countable ? count($result) : 0) === 0) {
                     // MX-record for domain can't be found
                     $this->returnStatus[] = Email::DNSWARN_NO_MX_RECORD;
                     $result = @dns_get_record($this->parseData[Email::COMPONENT_DOMAIN], DNS_A + DNS_CNAME);
-                    if (count($result) === 0) {
+                    if ((is_array($result) || $result instanceof Countable ? count($result) : 0) === 0) {
```
Comment (koriym) [inline src/Rule/Validate/Email.php]: dns_get_record never return  Countable?

File: src/Rule/Validate/Email.php [hunk 3]
```diff
@@ -65,110 +66,110 @@ class Email
...
 
     // Categories
-    const VALID_CATEGORY = 1;
-    const DNSWARN = 7;
-    const RFC5321 = 15;
-    const CFWS = 31;
-    const DEPREC = 63;
-    const RFC5322 = 127;
-    const ERR = 255;
+    public const VALID_CATEGORY = 1;
```
Comment (koriym) [inline src/Rule/Validate/Email.php]: Isn't "protected"?

Comment (koriym) [reply src/Rule/Validate/Email.php]: The const of this class is only used within this class.

File: src/Rule/AbstractCharCase.php [hunk 4]
```diff
@@ -68,7 +68,7 @@ protected function strtoupper($str)
      * @return int
```
Comment (koriym) [inline src/Rule/AbstractCharCase.php:68]:
```suggestion
     * @return int
```
The original declaration is wrong.

**Review by koriym — CHANGES_REQUESTED** [review]

File: src/Rule/AbstractCharCase.php (hunk 4 above)
Comment (koriym) [inline src/Rule/AbstractCharCase.php]:
```suggestion
```
The original type declaration is wrong.

**Review by koriym — CHANGES_REQUESTED** [review]

**Comment by harikt** [comment]
Thanks for your wonderful review . I will check when I get sometime

File: src/Rule/Validate/Email.php (hunk 1 above)
Comment (harikt) [reply src/Rule/Validate/Email.php]: preg_grep will be returning array, which means it is countable. But I agree we don't need to use Countable there. But preg_grep can also return false. So checking is_array seems a nice approach here. I will look more closely into this again.

**Review by koriym — APPROVED** [review]

File: src/Rule/Validate/Email.php (hunk 1 above)
Comment (harikt) [reply src/Rule/Validate/Email.php]: Hi @koriym , 

I am going to modify the code as 

```
} elseif (is_array(preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT)) && count(preg_grep('/^[0-9A-Fa-f]{0,4}$/', $matchesIP, PREG_GREP_INVERT)) !== 0) {
```

that will make sure there is no count(false)

**Comment by harikt** [comment]
Thank you @koriym . I have made the necessary changes you have requested for. In case you find any issues let me know.

**Pull request merged**